src/Access/AuthenticationData.cpp

```diff
@@ -105,12 +106,30 @@ AuthenticationData::Digest AuthenticationData::Util::encodeBcrypt(std::string_vi
 bool AuthenticationData::Util::checkPasswordBcrypt(std::string_view password [[maybe_unused]], const Digest & password_bcrypt [[maybe_unused]])
 {
 #if USE_BCRYPT
-    int ret = bcrypt_checkpw(password.data(), reinterpret_cast<const char *>(password_bcrypt.data()));  /// NOLINT(bugprone-suspicious-stringview-data-usage)
-    /// Before 24.6 we didn't validate hashes on creation, so it could be that the stored hash is invalid
-    /// and it could not be decoded by the library
-    if (ret == -1)
-        throw Exception(ErrorCodes::AUTHENTICATION_FAILED, "Internal failure decoding Bcrypt hash");
-    return (ret == 0);
+    /// Bcrypt takes a long time to compute, so we cache the results.
+    /// To avoid storing plaintext passwords in memory we only store SHA256 of the password from the user.
+    /// We store a mapping of the pair of SHA256 of the password and bcrypt hash to the result of the comparison.
+    using SimpleCacheBase = DB::CacheBase<std::string, bool>;
+    static auto bcrypt_cache = SimpleCacheBase("SLRU", CurrentMetrics::end(), CurrentMetrics::end(), /*max_size_in_bytes*/ 1024, /*max_count*/ 1024, /*size_ratio*/ 0.5);
+
+    auto password_digest = encodeSHA256(password);
+
+    std::string bcrypt_hash_str{password_bcrypt.data(), password_bcrypt.data() + password_bcrypt.size()};
+    std::string password_digest_str{password_digest.data(), password_digest.data() + password_digest.size()};
+
+    auto cache_key = password_digest_str + ":" + bcrypt_hash_str;
```

Comment:
maybe something like this to avoid string allocations
```
    auto cache_key = fmt::format(
        "{}:{}",
        std::string_view{reinterpret_cast<const char *>(password_digest.data()), password_digest.size()},
        std::string_view{reinterpret_cast<const char *>(password_bcrypt.data()), password_bcrypt.size()});
```

Reply:
Great idea!